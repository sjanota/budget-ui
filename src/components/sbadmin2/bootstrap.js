export const Variant = {
  secondary: 'secondary',
  primary: 'primary',
  danger: 'danger',
};

export const Size = {
  sm: 'sm',
  lg: 'lg',
};
