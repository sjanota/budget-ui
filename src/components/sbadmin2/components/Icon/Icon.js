import React from 'react';
import PropTypes from 'prop-types';

export default function Icon({ icon }) {
  return <i className={`fas fa-${icon} fa-fw`} />;
}

Icon.propTypes = {
  icon: PropTypes.string.isRequired,
};

Icon.Archive = 'archive';
Icon.Trash = 'trash-alt';
Icon.Edit = 'edit';
Icon.Plus = 'plus';
Icon.Minus = 'minus';
Icon.Save = 'save';
