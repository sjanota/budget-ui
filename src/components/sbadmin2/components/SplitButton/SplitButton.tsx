import { IconProp } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import classnames from 'classnames';
import React, { ButtonHTMLAttributes } from 'react';

import { Size, Variant } from '../../bootstrap.typed';

interface Props {
  icon: IconProp;
  variant: Variant;
  size?: Size;
  className?: string;
  disabled?: boolean;
  _ref?: { current: any };
  children?: string;
}

export default function SplitButton({
  icon,
  variant,
  size,
  className,
  disabled,
  _ref,
  children,
  ...props
}: Props & ButtonHTMLAttributes<HTMLButtonElement>) {
  const classNames = classnames('btn', 'btn-icon-split', className, {
    disabled,
    [`btn-${size}`]: size,
    [`btn-${variant}`]: variant,
  });

  return (
    <button
      className={classNames}
      disabled={disabled}
      ref={_ref}
      aria-label={children}
      {...props}
    >
      <span className='icon text-white-50'>
        <FontAwesomeIcon icon={icon} />
      </span>
      <span className='text'>{children}</span>
    </button>
  );
}
