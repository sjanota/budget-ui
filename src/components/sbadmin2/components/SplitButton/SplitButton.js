import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import Icon from '../Icon/Icon';

export default function SplitButton({
  children,
  icon,
  variant,
  size,
  className,
  disabled,
  ...props
}) {
  const classNames = classnames('btn', 'btn-icon-split', className, {
    disabled,
    [`btn-${size}`]: size,
    [`btn-${variant}`]: variant,
  });

  return (
    <button className={classNames} disabled={disabled} {...props}>
      <span className="icon text-white-50">
        <Icon icon={icon} />
      </span>
      <span className="text">{children}</span>
    </button>
  );
}

SplitButton.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  icon: PropTypes.string.isRequired,
  size: PropTypes.string,
  variant: PropTypes.string,
};
