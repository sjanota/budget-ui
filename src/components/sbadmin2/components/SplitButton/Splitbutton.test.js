import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import SplitButton from './SplitButton';
import Icon from '../Icon/Icon';
import { Variant, Size } from '../../bootstrap';

it('SplitButton renders with minimal props', () => {
  const component = renderer.create(<SplitButton icon={Icon.Archive} />);
  expect(component).toMatchSnapshot();
});

it('SplitButton renders with all props', () => {
  const component = renderer.create(
    <SplitButton
      icon={Icon.Archive}
      variant={Variant.danger}
      size={Size.lg}
      className="my-class"
      disabled
    >
      Text
    </SplitButton>
  );
  expect(component).toMatchSnapshot();
});

describe('SplitButton', () => {
  describe('when enabled', () => {
    const component = mount(
      <SplitButton
        icon={Icon.Archive}
        variant={Variant.danger}
        size={Size.lg}
        className="my-class"
      >
        Text
      </SplitButton>
    );
    it('passes icon to Icon', () => {
      expect(component.find(Icon)).toHaveProp('icon', Icon.Archive);
    });

    it('sets custom class', () => {
      expect(component.find('button')).toHaveClassName('my-class');
    });

    it('renders children', () => {
      expect(component.find('button')).toHaveText('Text');
    });

    it('sets proper btn-variant class', () => {
      expect(component.find('button')).toHaveClassName('btn-danger');
    });

    it('sets proper btn-size class', () => {
      expect(component.find('button')).toHaveClassName('btn-lg');
    });

    it('does not set disabled class', () => {
      expect(component.find('button')).not.toHaveClassName('disabled');
    });

    it('does not set disabled button prop', () => {
      expect(component.find('button')).toHaveProp('disabled', undefined);
    });
  });

  describe('when disabled', () => {
    const component = mount(<SplitButton icon={Icon.Archive} disabled />);

    it('sets disabled class', () => {
      expect(component.find('button')).toHaveClassName('disabled');
    });

    it('sets disabled button prop', () => {
      expect(component.find('button')).toHaveProp('disabled', true);
    });
  });
});
