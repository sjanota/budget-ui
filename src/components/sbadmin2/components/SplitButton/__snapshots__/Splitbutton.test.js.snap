// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplitButton renders with all props 1`] = `
<button
  className="btn btn-icon-split my-class disabled btn-lg btn-danger"
  disabled={true}
>
  <span
    className="icon text-white-50"
  >
    <i
      className="fas fa-archive fa-fw"
    />
  </span>
  <span
    className="text"
  >
    Text
  </span>
</button>
`;

exports[`SplitButton renders with minimal props 1`] = `
<button
  className="btn btn-icon-split"
>
  <span
    className="icon text-white-50"
  >
    <i
      className="fas fa-archive fa-fw"
    />
  </span>
  <span
    className="text"
  />
</button>
`;
